File: cheatsheet/编程笔记/_util/java/StringUtil.java
/**
 * <P> Title: 公用類別                                      </P>
 * <P> Description: 字串處理                                </P>
 * <P> Copyright: Copyright (c) 2010/07/31                  </P>
 * <P> Company:Everunion Tech. Ltd.                         </P>
 */

package com.util;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.ResultSet;
import java.util.Iterator;
import java.util.Map;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;

/**
 * 字串處理
 * @author Holer W. L. Feng
 * @version 0.1
 */
public final class StringUtil
{

    /**
     * 轉成字串
     * @param sour 待轉資料
     * @param init 預設值
     * @param length 截取字串長度,超過此長度者,後面顯示“...”;不希望截取可填入0
     * @return String 字串
     */
    public static String toString(Object sour, String init, int length)
    {
        // 初始化
        String dest = null;
        try
        {
            // 為空
            if ( sour == null || "".equals(sour) )
            {
                dest = init;
            }
            else
            {
                dest = "" + sour;
            }
        }
        // 拋出例外
        catch ( Exception e )
        {
            dest = init;
        }
        // 需要截取時(由於超出部分顯示“...”，所以必須3位以上才截取)
        if ( length > 0 && dest != null && dest.length() > length && dest.length() > 3 )
        {
            dest = dest.substring(0, length - 3) + "...";
        }
        return dest;
    }


    /**
     * 轉成字串
     * @param sour 待轉資料
     * @param init 預設值
     * @return String 字串
     */
    public static String toString(Object sour, String init)
    {
        return toString(sour, init, 0);
    }


    /**
     * 轉成字串
     * @param sour 待轉資料
     * @return String 字串
     */
    public static String toString(Object sour)
    {
        return toString(sour, "");
    }


    /**
     * 取得物件裏的值
     * @param sour 儲存著所需內容的物件，可以是 Map, HttpServletRequest, HttpSession, ResultSet 類別
     * @param name 對應的值的名稱，如果這個值為null，則認為 sour 本身是值
     * @param init 預設值，如果沒有取得對應值時，則傳回此值
     * @param length 截取字串長度,超過此長度者,後面顯示“...”;不希望截取可填入0
     * @return String 物件裏的值的字串
     */
    public static String getValue(Object sour, String name, String init, int length)
    {
        // 如果變數 sour 為空
        if ( sour == null )
        {
            return toString(init, init, length);
        }
        // 如果變數 name 為空，則認為 sour 本身是值
        if ( name == null )
        {
            return toString(sour, init, length);
        }

        // 根據 sour 的不同類型來取值
        try
        {
            // 如果 sour 是 String 類別的物件
            if ( sour instanceof String )
            {
                return toString(sour, init, length);
            }
            // 如果 sour 是 Map 類別的物件
            if ( sour instanceof Map )
            {
                Map<String, String> map = (Map)sour;
                // 如果此映射包含指定鍵的映射關系
                if ( map.containsKey(name) )
                {
                    return toString(map.get(name), init, length);
                }
                // 查詢資料庫時，鍵都會自動轉成小寫
                if ( map.containsKey(name.toLowerCase()) )
                {
                    return toString(map.get(name.toLowerCase()), init, length);
                }
                // 如果此映射不包含指定鍵的映射關系，則忽略大小寫來取值
                for ( Iterator<Map.Entry<String, String>> iter = map.entrySet().iterator(); iter.hasNext(); )
                {
                    Map.Entry<String, String> entry = iter.next();
                    // 如果鍵忽略大小寫時與 name 相同
                    if ( name.equalsIgnoreCase(toString(entry.getKey())) )
                    {
                        return toString(entry.getValue(), init, length);
                    }
                }
            }
            // 如果 sour 是 HttpServletRequest 類別的物件
            if ( sour instanceof HttpServletRequest )
            {
                // Parameter裡面有,則取Parameter的
                if ( ((HttpServletRequest)sour).getParameterMap().containsKey(name) )
                {
                    return toString(((HttpServletRequest)sour).getParameter(name), init, length);
                }
                // 如果 Parameter 不包含對應的名稱; 取Attribute的
                return toString(((HttpServletRequest)sour).getAttribute(name), init, length);
            }
            // 如果 sour 是 HttpSession 類別的物件
            if ( sour instanceof HttpSession )
            {
                return toString(((HttpSession)sour).getAttribute(name), init, length);
            }
            // 如果 sour 是 ResultSet 類別的物件
            if ( sour instanceof ResultSet )
            {
                return toString(((ResultSet)sour).getString(name), init, length);
            }
            // 如果 sour 不是上述類別的物件，則認為它自身是值。
            return toString(sour, init, length);
        }
        // 取值時出現例外
        catch ( Exception e )
        {
            return toString(init, init, length);
        }
    }


    /**
     * 取得物件裏的值
     * @param sour 儲存著所需內容的物件，可以是 Map, HttpServletRequest, HttpSession, ResultSet 類別
     * @param name 對應的值的名稱，如果這個值為空，則認為 sour 本身是值
     * @param init 預設值，如果沒有取得對應值時，則傳回此值
     * @return String 物件裏的值的字串
     */
    public static String getValue(Object sour, String name, String init)
    {
        return getValue(sour, name, init, 0);
    }


    /**
     * 取得物件裏的值的字串
     * @param sour 儲存著所需內容的物件，可以是 Map, HttpServletRequest, HttpSession, ResultSet 類別
     * @param name 對應的值的名稱，如果這個值為null，則認為 sour 本身是值
     * @return String 物件裏的值的字串。取不到值，或者值為 null 則傳回""
     */
    public static String getValue(Object sour, String name)
    {
        return getValue(sour, name, "");
    }


    /**
     * 取得物件裏的值的字串
     * @param sour 儲存著所需內容的物件，相當於 toString(sour)
     * @param name 對應的值的名稱，如果這個值為空，則認為 sour 本身是值
     * @return String 物件裏的值的字串。取不到值，或者值為 null 則傳回""
     */
    public static String getValue(Object sour)
    {
        return getValue(sour, null, "");
    }


    /**
     * 從右取資料
     * @param str 原資料
     * @param length 長度
     * @return String 子資料
     */
    public static String right(String str, int length)
    {
        int lengthstr = str.length();
        return str.substring(lengthstr - length);
    }


    /**
     * 轉成SQL字串
     * @param sour 待轉資料
     * @return String 字串
     */
    public static String toSqlStr(Object sour)
    {
        return toString(sour).replaceAll("'", "''");
    }


    /**
     * 轉成資料庫的比較字串
     * @param sour 待轉資料
     * @param toSqlStr 是否需要先轉成資料庫字串,預設為true
     * @return String 字串
     */
    public static String toEqualsSql(Object sour, boolean toSqlStr)
    {
        String retValue = toString(sour);
        // 轉成資料庫字串
        if ( toSqlStr )
            retValue = toSqlStr(sour);
        // 一個斜杠需轉換成兩個斜杠(MySQL 資料庫使用)
        return retValue.replaceAll("\\\\", "\\\\\\\\");
    }


    /**
     * 轉成資料庫的比較字串
     * @param sour 待轉資料
     * @return String 字串
     */
    public static String toEqualsSql(Object sour)
    {
        return toEqualsSql(sour, true);
    }


    /**
     * 轉成資料庫的like字串
     * @param sour 待轉資料
     * @param toSqlStr 是否需要先轉成資料庫字串,預設為true
     * @return String 字串
     */
    public static String toLikeSql(Object sour, boolean toSqlStr)
    {
        // 一個斜杠轉成四個(MySQL語法,由於前面已經將1個斜杠轉成兩個,這裡只需再轉成兩個就行了)
        String value = toEqualsSql(sour, toSqlStr).replaceAll("\\\\", "\\\\\\\\");
        // 處理特殊符號，這裡用“\”做轉義符號(“%”轉成“\%”,“_”轉成“\_”)；
        // 如果含有特殊符號，Oracle請在條件的最後面加上“escape'\'”
        value = value.replaceAll("%", "\\\\%").replaceAll("_", "\\\\_");
        // 將空格轉變成匹配任意內容
        return value.replaceAll("\\s", "%");
    }


    /**
     * 轉成資料庫的like字串
     * @param sour 待轉資料
     * @return String 字串
     */
    public static String toLikeSql(Object sour)
    {
        return toLikeSql(sour, true);
    }


    /**
     * 利用MD5進行加密,並用 16 進制的形式表示
     * @param source 待加密的字串
     * @return 加密後的字串
     */
    public static String MD5(String source)
    {
        char hexDigits[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
                'A', 'B', 'C', 'D', 'E', 'F'};
        try
        {
            byte[] strTemp = source.getBytes();
            // 確定加密方法
            MessageDigest mdTemp = MessageDigest.getInstance("MD5");
            mdTemp.update(strTemp);
            // MD5 的計算結果是一個 128 位的長整數, 用字元表示就是 16 個字節
            byte[] md = mdTemp.digest();
            int j = md.length;
            // 每個字節用 16 進制表示的話, 使用兩個字節, 所以表示成 16 進制需要 32 個字節
            char str[] = new char[j * 2];
            // 表示轉換結果中對應的字節位置
            int k = 0;
            // 逐個將 MD5 的字節轉換成 16 進制字節
            for ( int i = 0; i < j; i++ )
            {
                byte byte0 = md[i];
                // 取字節中高 4 位的數字轉換, >>> 為邏輯右移, 將符號位一起右移
                str[k++] = hexDigits[byte0 >>> 4 & 0xf];
                // 取字節中低 4 位的數字轉換
                str[k++] = hexDigits[byte0 & 0xf];
            }
            return new String(str);
        }
        // 捕獲例外
        catch ( NoSuchAlgorithmException e )
        {
            return "";
        }
    }
}

File: cheatsheet/编程笔记/_util/c#/_StringUtil.cs
/**
 * <P> Title: pili                          </P>
 * <P> Description: 公用函數                </P>
 * <P> Copyright: Copyright (c) 2009/4/2    </P>
 * <P> Company:Everunion Tech. Ltd.         </P>
 * @author Andy
 * @version 0.4 Original Design from design document.
 */
using System;
using System.Data;
using System.Collections;
using System.Configuration;
using System.Web;
using System.Web.Security;
using System.Text.RegularExpressions;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;
using System.Reflection;

using System.Text;


namespace Com.Everunion.Util
{
    /// <summary>
    /// 函數傳回值
    /// </summary>
    public class ReturnData
    {
        public const int Fail = -1;
        public const int Exists = 0;
        public const int Succ = 1;
    }


    /// <summary>
    /// 工具類
    /// </summary>
    public class StringUtil
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(typeof(StringUtil));
        /// <summary>
        /// 轉換成字串
        /// </summary>
        /// <param name="o">原資料</param>
        /// <param name="init">預設值</param>
        /// <returns> 轉換后的資料</returns>
        public static string ToStr(object o, String init)
        {
            string str = Convert.ToString(o);
            if (str != null && str.Length > 0)
                return str.Trim();
            //傳回預設
            else
                return init;
        }


        /// <summary>
        /// 傳回值轉換成訊息
        /// </summary>
        /// <param name="actionName">作業名稱(如：新增、修改等，預設為"作業")</param>
        /// <param name="returnValue">傳回值</param>
        /// <returns> 訊息</returns>
        public static string ReturnValue2Mess(string actionName, int returnValue)
        {
            actionName = String.IsNullOrEmpty(actionName) ? "作業" : actionName;
            //資料相同的訊息
            if ( ReturnData.Exists == returnValue )
                return "存在相同的資料," + actionName + "失敗!";
            //失敗訊息
            else if ( ReturnData.Fail == returnValue )
                return actionName + "失敗!";
            //成功訊息(新增時，傳回主鍵，所以會大於1)
            else if ( ReturnData.Succ <= returnValue )
                return actionName + "成功!";
            //無
            else
                return "ReturnData中無此訊息!";
        }


        /// <summary>
        /// 傳回值轉換成訊息
        /// </summary>
        /// <param name="returnValue">傳回值</param>
        /// <returns> 訊息</returns>
        public static string ReturnValue2Mess(int returnValue)
        {
            return ReturnValue2Mess("作業", returnValue);
        }


        /// <summary>
        /// 提示訊息
        /// </summary>
        /// <param name="response">回應物件</param>
        /// <param name="mess">訊息</param>
        /// <returns> 提示訊息</returns>
        public static void AlertMessage(string mess)
        {
            mess = mess.Replace("'", "&#39;");
			callJsFunction("alert('" + mess + "!');");
        }


        /// <summary>
        /// 提示訊息
        /// </summary>
        /// <param name="actionName">作業名稱(如：新增、修改、刪除等，預設為"作業")</param>
        /// <param name="dataName">資料主鍵名稱</param>
        /// <param name="rd">結果狀態</param>
        /// <returns> 提示訊息</returns>
        public static void AlertMessage(string actionName, string dataName, int rd)
        {
            string mess = "";
            actionName = String.IsNullOrEmpty(actionName) ? "作業" : actionName;
            dataName = String.IsNullOrEmpty(dataName) ? "資料" : dataName;
            //主鍵重複
            if ( ReturnData.Exists == rd )
            {
                mess = "已經存在相同的 " + dataName + ", " + actionName + "失敗";
            }
            //操失敗
            else if ( ReturnData.Fail == rd )
            {
                mess = actionName + "失敗";
            }
            //操作成功(新增時，傳回主鍵，所以會大於1)
            else if ( ReturnData.Succ <= rd )
            {
                mess = actionName + "成功";
            }
            AlertMessage(mess);
        }



        /// <summary>
        /// 提示訊息
        /// </summary>
        /// <param name="dataName">資料主鍵名稱</param>
        /// <param name="rd">結果狀態</param>
        /// <returns> 提示訊息</returns>
        public static void AlertMessage(string dataName, int rd)
        {
            AlertMessage(null, dataName, rd);
        }


        /// <summary>
        /// 執行js函數
        /// </summary>
        /// <param name="response">回應物件</param>
        /// <param name="mess">訊息</param>
        public static void callJsFunction(string jsFunction)
        {
            HttpResponse response = HttpContext.Current.Response;
            response.Write("<script defer language='javascript' type='text/javascript'>" + jsFunction + "</script>");
        }


        /// <summary>
        /// 轉從合法的SQLCommand
        /// </summary>
        /// <param name="o">原字串</param>
        /// <returns>SQLCommand</returns>
        public static string ToSqlStr(String o)
        {
            string str = ToStr(o);
            return str.Replace("'", "''");
        }


        /// <summary>
        /// 過濾模糊查詢關鍵字符，并過濾單引號
        /// </summary>
        /// <param name="o">次序參數</param>
        /// <returns>處理了'%','['等</returns>
        public static string ToLikeSqlStr(String o)
        {
            return ToLikeSqlStr(o, true);
        }


        /// <summary>
        /// 轉成合法的SQLCommand
        /// </summary>
        /// <param name="o">原字串</param>
        /// <param name="filterSql">為True時過濾單引號（'）</param>
        /// <returns>SQLCommand</returns>
        public static string ToLikeSqlStr(String o, bool filterSql)
        {
            string str;
            //過濾＇
            if (filterSql)
            {
                str = ToSqlStr(o);
            }
            //不過濾＇
            else
            {
                str = ToStr(o);
            }
            //MySQL
            if ( HibernateHelper.Dialect.GetType().FullName.IndexOf("MySQLDialect") > -1 )
            {
                return str.Replace("%", @"\%").Replace("_", @"\_").Replace("[", @"\[");
            }
            //SqlServer2000
            else if ( HibernateHelper.Dialect.GetType().FullName.IndexOf("MsSql2000Dialect") > -1 )
            {
                return str.Replace("%", @"[%]").Replace("_", @"[_]").Replace("[", @"[[]");
            }
            //其他，待完成
            else
            {
                return str.Replace("%", "").Replace("_", @"").Replace("[", @"");
            }
        }


        /// <summary>
        /// 設定表單資料
        /// </summary>
        /// <param name="paramObj">參數物件</param>
        /// <param name="destObj">表單物件</param>
        public static void SetPageValue(Object paramObj, Object destObj)
        {
            //無資料
            if (paramObj == null)
            {
                return;
            }
            //屬性
            PropertyInfo[] propertys = paramObj.GetType().GetProperties();
            //取參數
            foreach (PropertyInfo p in propertys)
            {
                string name = p.Name.ToString();
                FieldInfo destField = destObj.GetType().GetField(name, BindingFlags.Instance | BindingFlags.NonPublic);
                if ( destField != null )
                {
                    PropertyInfo destPro1 = destField.FieldType.GetProperty("SelectedValue");
                    PropertyInfo destPro2 = destField.FieldType.GetProperty("Text");
                    PropertyInfo destPro3 = destField.FieldType.GetProperty("Value");
                    //radio或checkbox
                    if ( destPro1 != null )
                    {
                        destPro1.SetValue(destField.GetValue(destObj), ToStr(p.GetValue(paramObj, null)), null);
                    }
                    //text
                    else if ( destPro2 != null )
                    {
                        destPro2.SetValue(destField.GetValue(destObj), ToStr(p.GetValue(paramObj, null)), null);
                    }
                    //hidden
                    else if ( destPro3 != null )
                    {
                        destPro3.SetValue(destField.GetValue(destObj), ToStr(p.GetValue(paramObj, null)), null);
                    }
                }
            }
        }


        /// <summary>
        /// 參數設定
        /// </summary>
        /// <typeparam name="T">傳回的類型</typeparam>
        /// <param name="type">實例類型</param>
        /// <param name="request">HTTP伺服器請求</param>
        /// <returns>該類型實例</returns>"Id:WareId","Oid:id"
        public static T GetInstanceByParam<T>(Type type, params String[] aliasArray)
        {
            if ( HttpContext.Current == null )
                throw new Exception("非http請求不可呼叫此函數");
            //別名
            Hashtable ht = new Hashtable();
            if ( aliasArray != null && aliasArray.Length > 0 )
            {
                //迭代別名
                foreach ( string alias in aliasArray )
                {
                    String[] al = alias.Split(':');
                    ht.Add(al[0], al[1]);
                }
            }
            //HTTP伺服器請求
            HttpRequest request = HttpContext.Current.Request;
            object obj = Activator.CreateInstance(type);
            PropertyInfo[] prop = type.GetProperties();
            //參數

            foreach ( PropertyInfo pi in prop )
            {
                if ( pi.CanWrite )
                {
                    try
                    {
                        switch ( pi.PropertyType.ToString() )
                        {
                            case "System.SByte":
                                pi.SetValue(obj, Convert.ToSByte(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            case "System.Byte":
                                pi.SetValue(obj, Convert.ToByte(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            case "System.String":
                                pi.SetValue(obj, StringUtil.ToStr(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            //時間
                            case "System.DateTime":
                                pi.SetValue(obj, StringUtil.ToDateForStr(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            case "System.Int16":
                                pi.SetValue(obj, StringUtil.ToShort(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            case "System.Int32":
                                pi.SetValue(obj, StringUtil.ToInt(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            //長整形
                            case "System.Int64":
                                pi.SetValue(obj, StringUtil.ToLong(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            case "System.Single":
                                pi.SetValue(obj, StringUtil.toFloat(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            case "System.Double":
                                pi.SetValue(obj, StringUtil.toDouble(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            //decimal
                            case "System.Decimal":
                                pi.SetValue(obj, StringUtil.toDecimal(request.Form[StringUtil.ToStr(ht[pi.Name], pi.Name)]), null);
                                break;
                            default:
                                log.Error("沒有取到參數:" + pi.PropertyType.ToString());
                                break;
                        }
                    }
                    catch
                    {
                    }
                }
            }
            //傳回
            return (T)obj;
        }


        /// <summary>
        /// 轉換成字串
        /// </summary>
        /// <param name="o">原資料</param>
        /// <returns> 轉換后的資料</returns>
        public static string ToStr(object o)
        {
            return ToStr(o, "");
        }


        /// <summary>
        /// 轉成int
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <returns> 轉換后的資料</returns>
        public static int ToInt(object o)
        {
            return ToInt(o, 0);
        }


        // <summary>
        /// 轉成短整形
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <returns> 轉換后的資料</returns>
        public static short ToShort(object o)
        {
            try
            {
                //轉換 
                return Convert.ToInt16(ToStr(o, "0"));
            }
            catch
            {
                //格式錯誤
                return 0;
            }
        }


        // <summary>
        /// 轉成長整形
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <returns> 轉換后的資料</returns>
        public static long ToLong(object o)
        {
            try
            {
                //轉換 
                return Convert.ToInt64(ToStr(o, "0"));
            }
            catch
            {
                //格式錯誤
                return 0;
            }
        }


        // <summary>
        /// 轉成單精度
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <returns> 轉換后的資料</returns>
        public static float toFloat(object o)
        {
            try
            {
                //轉換 
                return Convert.ToSingle(ToStr(o, "0"));
            }
            catch
            {
                //格式錯誤
                return 0;
            }
        }


        // <summary>
        /// 轉成雙精度
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <returns> 轉換后的資料</returns>
        public static double toDouble(object o)
        {
            try
            {
                //轉換 
                return Convert.ToDouble(ToStr(o, "0"));
            }
            catch
            {
                //格式錯誤
                return 0;
            }
        }


        /// <summary>
        /// 轉成Decimal型
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <returns>轉換后的資料</returns>
        public static decimal toDecimal(object o)
        {
            try
            {
                //轉換
                return Convert.ToDecimal(ToStr(o, "0"));
            }
            catch
            {
                //格式錯誤
                return 0;
            }
        }

        /// <summary>
        /// 轉成int
        /// </summary>
        /// <param name="o">待轉換之資料</param>
        /// <param name="init">預設資料</param>
        /// <returns> 轉換后的資料</returns>
        public static int ToInt(object o, int init)
        {
            try
            {
                //轉換 
                return Convert.ToInt32(ToStr(o, "" + init));
            }
            catch
            {
                //格式錯誤
                return init;
            }

        }


        /// <summary>
        /// 判断存在
        /// </summary>
        /// <param name="o">待驗證的參數</param>
        /// <returns>有資料傳回TRUE</returns>
        public static bool Requied(object o)
        {
            string str = ToStr(o);
            //有資料
            if (str != null && str.Length > 0)
                return true;
            else
                return false;
        }


        /// <summary>
        /// 格式化時間到字串
        /// </summary>
        /// <param name="date">待轉時間</param>
        /// <param name="format">時間格式</param>
        /// <returns>格式化后的字串</returns>
        public static string ToStrForDate(string date, string format)
        {
            return ToStrForDate(DateTime.Parse(date), format);
        }


        /// <summary>
        /// 格式化時間到字串
        /// </summary>
        /// <param name="date">待轉時間</param>
        /// <param name="format">時間格式</param>
        /// <returns>格式化后的字串</returns>
        public static string ToStrForDate(DateTime date, string format)
        {
            format = ToStr(format, "d");
            return string.Format("{0:" + format + "}", date);
        }



        /// <summary>
        /// 格式化時間到字串
        /// </summary>
        /// <param name="dateStr">待轉時間</param>
        /// <returns>格式化后的字串</returns>
        public static DateTime ToDateForStr(string dateStr)
        {
            return Convert.ToDateTime(dateStr);
        }


        /// <summary>
        /// 格式納秒時間到字串
        /// </summary>
        /// <param name="ticks">納秒</param>
        /// <param name="format">時間格式</param>
        /// <returns>格式化后的字串<</returns>
        public static string ToLongForDate(long ticks, string format)
        {
            return (new DateTime(ticks)).ToString(format);
        }


        /// <summary>
        /// 格式納秒時間到字串
        /// </summary>
        /// <param name="ticks">納秒</param>
        /// <returns>格式化后的字串<</returns>
        public static string ToLongForDate(long ticks)
        {
            return ToLongForDate(ticks, "yyyy/MM/dd hh:mm");
        }


        /// <summary>
        /// 格式納秒時間到字串
        /// </summary>
        /// <param name="ticks">納秒字串</param>
        /// <returns>格式化后的字串<</returns>
        public static string ToLongForDate(string ticks)
        {
            return ToLongForDate(ToLong(ticks), "yyyy/MM/dd hh:mm");
        }



        /// <summary>
        /// MD5加密
        /// </summary>
        /// <param name="o">原資料</param>
        /// <returns>加密結果</returns>
        public static string ToStrByMD5(object o)
        {
            string str = ToStr(o);
            //不為空才加密
            if ( str.Length > 0 )
                return System.Web.Security.FormsAuthentication.
                    HashPasswordForStoringInConfigFile(str, "MD5");
            else
                return "";
        }


        /// <summary>
        /// 取得時間差
        /// </summary>
        /// <param name="begin">開始納秒</param>
        /// <param name="end">結束納秒</param>
        /// <returns>格式化后的字串</returns>
        public static string ToTimeSpan(long begin, long end)
        {
            //操作訊息
            string recode = "";
            TimeSpan ts = new TimeSpan(end - begin);
            //天
            if ( ts.Days > 0 )
            {
                recode += ts.Days + "天 ";
            }
            //小時
            if ( ts.Hours > 0 )
            {
                recode += ts.Hours + "小時 ";
            }
            //分
            if ( ts.Minutes > 0 )
            {
                recode += ts.Minutes + "分 ";
            }
            //秒
            if ( ts.Seconds > 0 )
            {
                recode += ts.Seconds + "秒 ";
            }
            //操作訊息
            return recode;
        }


        /// <summary>
        /// 取得時間差
        /// </summary>
        /// <param name="begin">開始納秒字串</param>
        /// <param name="end">結束納秒字串</param>
        /// <returns>格式化后的字串</returns>
        public static string ToTimeSpan(string begin, string end)
        {
            return ToTimeSpan(ToLong(begin), ToLong(end));
        }


        /// <summary>
        /// 驗證EMAIL
        /// </summary>
        /// <param name="email">待驗證的EMAIL</param>
        /// <returns>合法傳回true</returns>
        public static bool ValidateEmail(string email)
        {
            Match m = Regex.Match(email, "\\w+([-+.\']\\w+)*@\\w+([-.]\\w+)*\\.\\w+([-.]\\w+)*");
            //傳回結果
            return m.Success;
        }


        /// <summary>
        /// 日期轉納秒字串
        /// </summary>
        /// <param name="date">日期(yyyy/mm/dd)</param>
        /// <returns>納秒字串</returns>
        public static long ToTicks(string date)
        {
            DateTime dt = Convert.ToDateTime(date);
            return dt.Ticks;
        }


        /// <summary>
        /// 驗證 Sql Injection
        /// </summary>
        /// <param name="inputs">輸入字符</param>
        /// <returns>True表示存在關鍵字</returns>
        public static bool SqlValidate(string inputs)
        {
            /*
           "要求驗證偵測到具有潛在危險的用戶端輸入值，對這個要求的處理已經中止。", _
           "具有潛在危險 "& RequestType &" 的值已從用戶端"&key&"="& Server.UrlEncode(Collection(key)) &"偵測到。" & _
           "這個值可能表示有人嘗試危害應用程式的安全性，例如SQL的注入攻擊。我們強烈建議您的應用程式應該明確地檢查所有這類的輸入。"
            */
            //參照include/module/config.asp
            string pattern = @"\s*select\s+|\s*update\s+|\s*insert\s+|\s*delete\s+|\s*drop\s+|and\s+\d+=\d+|\s*sp_|\s*declare\s+|\s*(Exec|Char|Cast|Convert)\s*\(";
            if ( Regex.Match(inputs, pattern, RegexOptions.IgnoreCase).Success )
                return true;
            else
                return false;
        }


        /// <summary>
        /// 過濾html標簽和JS
        /// </summary>
        /// <param name="str">待過濾資料</param>
        /// <returns>編碼后的結果</returns>
        public static String ToHtml(string str)
        {
            //該表達式待改進";
            string pattern = @"<([\s]*?/?[\s]*?(meta|iframe|script|frame|style|embed|object)[\s\S]*?/?[\s]*?)>";
            string destination = "&lt;$1&gt;";
            return Regex.Replace(Regex.Replace(Regex.Replace(str, @"(>?[^<]*?)>", "$1&gt;"), @"<([^>]*?<?)", "&lt;$1"), pattern, destination, RegexOptions.IgnoreCase);
            //return Regex.Replace(str, pattern, destination, RegexOptions.IgnoreCase);
        }



        /// <summary>
        /// 取得流水號前的日期字串
        /// </summary>
        /// <returns></returns>
        public static String GetNowaDateInSerial()
        {
            return DateTime.Now.ToString("yyMMdd");
        }
    }
}
